.github/workflows/create-design-review-issue.js

```diff
@@ -233,24 +305,107 @@ function env(key) {
+      issue,
+    }))
+    .sort((a, b) => a.sessionDate - b.sessionDate);
+}
+
+async function closeStaleIssues(token, repo, issuesWithSessionDate) {
+  const now = new Date();
+
+  // Compensate duration so that we swap only once the session has ended.
+  now.setHours(now.getHours() - sessionDurationHours);
+
+  const issues = issuesWithSessionDate.filter(
+    ({sessionDate}) => sessionDate < now
+  );
+  for (const {issue} of issues) {
+    const {number, title} = issue;
+    if (!isDryRun) {
+      await unpinGithubIssue(token, repo, number);
+      await closeGithubIssue(token, repo, number);
+    }
+    console./*OK*/ log('Unpinned & closed: ', title, '\n');
+  }
+  return issues;
+}
+
+async function closeStalePinNextIssue(token, repo, existing) {
+  const staleIssues = closeStaleIssues(token, repo, existing);
+  if (!staleIssues.length) {
+    // If there aren't any open stale issues, the newer issue has been pinned.
+    console./*OK*/ log('(Zero issues to close, pin or unpin.)');
+    return;
+  }
+
+  const mostRecentStaleIssue = staleIssues[staleIssues.length - 1];
+  const nextIssue = existing.find(
+    ({sessionDate}) => sessionDate > mostRecentStaleIssue.sessionDate
+  );
+  if (!nextIssue) {
+    throw new Error(
+      "Could not find next session issue to pin. If it's created later, it will NOT be pinned."
+    );
+  }
+
+  const {number, title} = nextIssue.issue;
+  if (!isDryRun) {
+    await pinGithubIssue(token, repo, number);
+  }
+  console./*OK*/ log('Pinned: ', title, '\n');
+}
+
+async function createNextDesignReviewIssue(token, repo, existing) {
   const nextIssueData = getNextIssueData();
+  const nextIssueDateFromTitle = getSessionDateFromTitle(nextIssueData.title);
 
-  if (process.argv.includes('--dry-run')) {
-    console./*OK*/ log(nextIssueData);
+  const existingIssue = existing.find(
+    ({sessionDate}) => sessionDate == nextIssueDateFromTitle
+  );
+  if (existingIssue) {
+    const {title, 'html_url': htmlUrl} = existingIssue.issue;
+    console./*OK*/ log(
+      '(Skipping creation of next issue since it exists.)\n' +
+        `- ${title}\n  ${htmlUrl}`
+    );
     return;
   }
 
+  if (isDryRun) {
+    console./*OK*/ log(nextIssueData);
+    return;
+  }
   const {title, 'html_url': htmlUrl} = await postGithubIssue(
-    env('GITHUB_TOKEN'),
-    env('GITHUB_REPOSITORY'),
+    token,
+    repo,
     nextIssueData
   );
   console./*OK*/ log(title);
   console./*OK*/ log(htmlUrl);
 }
 
-createDesignReviewIssue().catch((e) => {
-  console./*OK*/ error(e);
-  process.exit(1);
-});
+async function createDesignReviewIssue(token, repo) {
+  const existing = await getExistingIssuesWithSessionDate(token, repo);
+  await createNextDesignReviewIssue(token, repo, existing);
+  await closeStalePinNextIssue(token, repo, existing);
+}
+
+createDesignReviewIssue(env('GITHUB_TOKEN'), env('GITHUB_REPOSITORY')).catch(
+  (e) => {
+    console./*OK*/ error(e);
+    process.exit(1);
+  }
```

Comment:
I'd imagine this error handling will automatically happen? If not, maybe the logic should be explicitly co-located with the code that could possibly throw an error?

Reply:
It's to handle top-level `await` without an `UnhandledPromiseRejection`.